FEATURE: Chat channel preview card

martin-brennan opened the pull request:
This commit adds a preview card for a channel that the current user
is previewing when they don't have a membership for the channel.
It shows the channel title + badge based on the category, and allows
the user to join the channel as well as return to the channel browse
list.

**Desktop - With Description**

<img src="https://user-images.githubusercontent.com/920448/178172963-99dd4855-8ad3-49e0-a640-8960b841f1ec.png" width="300">

**Desktop - No Description**

<img src="https://user-images.githubusercontent.com/920448/178172935-ca30c270-f599-4841-a383-c0503b11977b.png" width="300">

**Mobile - With Description**

<img src="https://user-images.githubusercontent.com/920448/178172852-b99f737f-4df9-45cb-8974-d0ca3ff04e9d.png" width="300">

**Mobile - No Description**

<img src="https://user-images.githubusercontent.com/920448/178172868-8c707fb6-ea36-4351-90b8-1a7fa5d8baf3.png" width="300">

jjaffeux commented on assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:
```diff
@@ -0,0 +1,12 @@
+    class="chat-channel-preview-card__join-channel-button btn-primary"
+  }}
+  <p class="chat-channel-preview-card__browse-all">
+    <a href="/chat/browse">{{i18n "chat.browse_all_channels"}}</a>
```
Maybe use a link-to here

jjaffeux commented on assets/stylesheets/common/chat-channel-preview-card.scss:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  &__browse-all {
+    font-size: .875rem;
```
Cant we use a var here?

jjaffeux commented on assets/stylesheets/common/chat-channel-preview-card.scss:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  &__join-channel-button {
+    font-size: 1.25rem;
```
same cant we use a var?

jjaffeux commented on assets/stylesheets/common/chat-channel-preview-card.scss:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  .chat-channel-title__name {
+    font-size: 1.25rem;
```
var :p

jjaffeux commented on assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:
```diff
@@ -0,0 +1,12 @@
+  {{d-button
+    action=(action "onJoinChannel")
+    label="chat.channel_settings.join_channel"
+    class="chat-channel-preview-card__join-channel-button btn-primary"
```
we can go with -btn instead of -button

jjaffeux commented on assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:
```diff
@@ -0,0 +1,12 @@
+<div class="chat-channel-preview-card">
+  {{chat-channel-title channel=chatChannel}}
+  <p class="chat-channel-preview-card__description">{{or chatChannel.description "-"}}</p>
```
do we want to show it at all if empty?

jjaffeux reviewed (commented):
Sorry sa after it was a draft

jjaffeux commented:
@martin-brennan as you will have to write tests for this, please have a look to what I did here with the help of @cvx: https://github.com/discourse/discourse-chat/commit/458f25368bc2d80c385ce010b958a6ae4c56fe17#diff-d2d0cde095125acf48a06ed57196e54d70385e056b753047c92eac44c112b4e6R112

We can finally use new latest ember test syntax and I would want us to use it for chat as much as possible.

martin-brennan commented on assets/javascripts/discourse/components/chat-channel-about-view.js:
```diff
@@ -19,8 +19,13 @@ export default class ChatChannelAboutView extends Component {
       .then((membership) => {
-        this.channel.set("following", true);
-        this.channel.set("memberships_count", membership.user_count);
+        this.channel.setProperties({
```
This is now done in a few places in the app, maybe we can move it into the `followChatChannel` function, passing the whole channel object instead of just the ID?

martin-brennan commented:
@jjaffeux can you please take a look at this one now that it's ready?

martin-brennan commented:
Only test failures are to do with quoting, not sure why:

```
Error: QUnit Test Failure: Global error: Uncaught Error: Assertion Failed: expected container not to be destroyed at http://localhost:7357/assets/vendor.js, line 38378
Testem finished with non-zero exit code. Tests failed.
 While executing test: Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
not ok 267 Chrome 103.0 - [undefined ms] - Global error: Uncaught Error: Assertion Failed: expected container not to be destroyed at http://localhost:7357/assets/vendor.js, line 38378
 While executing test: Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
    ---
        browser log: |
            {"type":"error","text":"Uncaught Error: Assertion Failed: expected container not to be destroyed at http://localhost:7[357](https://github.com/discourse/discourse-chat/runs/7275262839?check_suite_focus=true#step:23:358)/assets/vendor.js, line 38378\n","testContext":{"id":6,"name":"Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in","items":[],"state":"complete"}}
    ...
Error: QUnit Test Failure: Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
not ok 268 Chrome 103.0 - [562 ms] - Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
    ---
        actual: >
            /chat/channel/7/Uncategorized
        expected: >
            /c/uncategorized/1
        stack: >
                at Object.eval (acceptance/chat-quoting-test:117:14)
        message: >
            navigates to the chatable url
        negative: >
            false
        browser log: |
    ...
error Command failed with exit code 1.
```

martin-brennan commented:
Had a similar error to the quoting one locally, not sure what `global failure: Error` means:

```
not ok 67 Chrome 100.0 - [120049 ms] - Discourse Chat | Component | chat-user-avatar: user is online
    ---
        stack: >
                at http://localhost:7357/assets/chunk.vendors-node_modules_pretender_dist_pretender_es_js-node_modules_qunit_qunit_qunit_js-node_mo-a05e2f.453a7e9daf9c124ace08.js:5046:251
        message: >
            global failure: Error
        negative: >
            false
        browser log: |
    ...
```

The tests did not even finish running locally...

martin-brennan commented:
Okay deleting `test/javascripts/components/chat-channel-preview-card-test.js`, the new test I added, locally makes the suite finish locally, but one of the quoting tests still fails:

```
not ok 264 Chrome 100.0 - [499 ms] - Acceptance: Discourse Chat | quoting on mobile: it opens the chatable, opens the composer, and pastes the markdown in
    ---
        actual: >
            /chat/channel/7/Uncategorized
        expected: >
            /c/uncategorized/1
        stack: >
                at Object.eval (acceptance/chat-quoting-test:117:14)
        message: >
            navigates to the chatable url
        negative: >
            false
        browser log: |
    ...
Testem finished with non-zero exit code. Tests failed.
```

@jjaffeux can you please look at the test I added in this PR and see if there is anything that jumps out as wrong with it? It seems so simple IDK why it would mess things up...

jjaffeux commented:
@Chapoi maybe we limit padding on mobile screens, Im afraid it might take too much space on small screens?

Chapoi commented on assets/javascripts/discourse/templates/components/chat-channel-preview-card.hbs:
```diff
@@ -0,0 +1,16 @@
+    label="chat.channel_settings.join_channel"
+    class="chat-channel-preview-card__join-channel-btn btn-primary"
+  }}
+  <p class="chat-channel-preview-card__browse-all">
```
Why is this wrapped in a p-tag? It's semantically not correct and it adds margin we don't want.
I would prefer the a-tag directly, without a wrapping element, but if for some ember-reason you need one, I'd prefer a simple span.

martin-brennan replied:
No reason particularly, will change :)

martin-brennan replied:
There is some margin/padding in the design that is lost if I remove the P tag though:

![image](https://user-images.githubusercontent.com/920448/178207170-6d1ad420-f7d0-459a-99a8-af4840a7beb0.png)

Here is what it looks like locally now:

![image](https://user-images.githubusercontent.com/920448/178207209-96fa3af1-cd2f-4ad1-829c-932016ab0dbf.png)

If you want you can do another round of tweaks and push to this PR? I will just remove the P tag

Chapoi commented:
> @Chapoi maybe we limit padding on mobile screens, Im afraid it might take too much space on small screens?

hm, it's looking fine to me like this
<img width="393" alt="image" src="https://user-images.githubusercontent.com/101828855/178209679-9808fda1-1013-424a-af44-b7e9dc86143e.png">

And since it's literally the main thing going on on the page, I think it's okay to take up this much space.

jjaffeux commented on assets/javascripts/discourse/components/chat-channel-preview-card.js:
```diff
@@ -0,0 +1,45 @@
+    return !isEmpty(this.channel.description);
+  }
+
+  @computed("channel.description")
```
you can compute directly on hasDecription here

jjaffeux commented on test/javascripts/acceptance/chat-test.js:
```diff
@@ -1334,15 +1326,6 @@ acceptance(
       });
     });
 
-    test("Composer placeholder is specific when previewing", async function (assert) {
```
Should we remove chat.placeholder_previewing key now ?

jjaffeux commented on test/javascripts/components/chat-channel-preview-card-test.js:
```diff
@@ -0,0 +1,87 @@
+import { setupRenderingTest } from "discourse/tests/helpers/component-test";
```
we should have a test for onJoinChannel

martin-brennan replied:
I am not sure the best way to test this for a component test? The ember guide mentions using service stubs which we do have for chat https://guides.emberjs.com/release/testing/testing-components/ but I'm not sure what exactly I should be testing wrt AJAX calls and things like that. The chat-browse-test already checks that the join button works and the card is gone / channel is reloaded. I am going to merge this PR, but would like your opinion on this and I can add another test afterward if needed.

jjaffeux replied:
I think what we want to test here os that clicking the button calls an action named onJoinChannel

martin-brennan replied:
There doesn't really seem to be an easy way to do that though? In https://guides.emberjs.com/release/testing/testing-components/#toc_testing-actions they presume that the action is being passed in via args like this:

```
export default class CommentFormComponent extends Component {
  @tracked comment = '';

  @action
  submitComment() {
    this.args.submitComment({ comment: this.comment });
  }
}
```

```
<CommentForm @submitComment={{this.externalAction}} />
```

Which is not the case for the preview card. I've tried a few different ways of overriding `onJoinChannel` in the test with a stub and none of them worked. And anyway not sure what value testing it here adds when we already test clicking the button in chat-browse-test? If you had an example of what you mean that would be great.

jjaffeux replied:
I looked at this today and made it work but I want to write more dedicated guidelines for this. Im just like you and I'm learning to adapt to our new patterns and also improve what we currently do.

I think the good separation is:
- one happy path acceptance test
- lots of component/unit tests for various edge cases

I will elaborate more on this in the near future.

martin-brennan replied:
Nice, thank you for looking into this!

jjaffeux commented on test/javascripts/acceptance/chat-browse-test.js:
```diff
@@ -77,6 +106,37 @@ acceptance("Discourse Chat - chat browsing", function (needs) {
+    await click(".chat-channel-unfollow");
+    await click(".chat-channel-preview");
+
+    assert.ok(
```
I think we could get away with just this test here, we are actually kind duplicating tests here, we want to ensure the card is showing, we already have various tests to assert the card is acting as we want it to

jjaffeux commented on assets/javascripts/discourse/templates/components/chat-live-pane.hbs:
```diff
@@ -1,5 +1,5 @@
 {{#if (and fullPage includeHeader)}}
-  <div class="chat-full-page-header {{if previewing "previewing"}}">
+  <div class="chat-full-page-header {{if previewing 'previewing'}}">
```
For now please fix it manually @martin-brennan, it's on @cvx plate to get linting on par between core and chat

jjaffeux commented:
Tried on small screen and dark mode, looks great :heart:

<img width="386" alt="Screenshot 2022-07-11 at 12 28 56" src="https://user-images.githubusercontent.com/339945/178244937-b383d2e3-1db0-42f4-be78-35ea7b47be20.png">

jjaffeux commented:
ok tests should be fixed here, I don't have an exact answer on "why" for everything yet, I'm new to this new style too :p But I think it's good enough to move on.

@martin-brennan @Chapoi great work :+1:

jjaffeux reviewed (approved):
Just be sure to fix linting first
